File: backend/app/routers/me.py
"""GET /api/me — the protected endpoint that returns the authenticated user.

Auth is enforced by ``get_current_claims`` (see ``deps.py``), which
extracts and verifies the JWT from the HttpOnly cookie. By the time
the handler runs, the request is guaranteed authenticated.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status

from app.deps import get_current_claims, get_store
from app.schemas import UserView
from app.services.token import Claims
from app.store import Store, UserNotFoundError

router = APIRouter(tags=["me"])


@router.get("/me", response_model=UserView)
def get_me(
    claims: Claims = Depends(get_current_claims),
    store: Store = Depends(get_store),
) -> UserView:
    """Re-fetch from the DB rather than trusting only the token.

    This means a deleted user invalidates immediately on the next /me
    call even though the JWT is still cryptographically valid — a small
    consolation against the lack of a revocation list (see REPORT §2).
    """
    try:
        user = store.user_by_id(claims.user_id)
    except UserNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail={"error": "authentication required", "code": "unauthenticated"},
        )
    return UserView(id=user.id, email=user.email)

File: backend/internal/handlers/me.go
package handlers

import (
	"net/http"

	"github.com/chaninjung/minimal-auth-app/internal/httpx"
	"github.com/chaninjung/minimal-auth-app/internal/middleware"
	"github.com/chaninjung/minimal-auth-app/internal/store"
)

// Me handles GET /api/me — the protected endpoint that returns the
// authenticated user. Auth is enforced by middleware.Auth which sets the
// claims on the request context.
type Me struct {
	Store *store.Store
}

func (h *Me) Get(w http.ResponseWriter, r *http.Request) {
	claims := middleware.ClaimsFrom(r.Context())
	if claims == nil {
		// Defensive — should be unreachable behind middleware.Auth.
		httpx.Error(w, http.StatusUnauthorized, "unauthenticated", "authentication required")
		return
	}
	// Re-fetch from the DB rather than trusting only the token. This means
	// if the user is deleted, the next /me call invalidates the session
	// even though the JWT itself is still cryptographically valid.
	u, err := h.Store.UserByID(r.Context(), claims.UserID)
	if err != nil {
		httpx.Error(w, http.StatusUnauthorized, "unauthenticated", "authentication required")
		return
	}
	httpx.JSON(w, http.StatusOK, userView{ID: u.ID, Email: u.Email})
}
